noxfile.py

```diff
@@ -0,0 +1,105 @@
+import nox
+import os
+import sys
+
+
+@nox.session(name="pybamm-requires")
+def run_pybamm_requires(session):
+    if sys.platform != "win32" or sys.platform != "darwin":
+        session.install("wget", "cmake")
+        session.run("python", "scripts/install_KLU_Sundials.py")
+        session.run(
+            "git",
+            "clone",
+            "https://github.com/pybind/pybind11.git",
+            "pybind11/",
+            external=True,
+        )
+
+
+@nox.session(name="coverage")
+def run_coverage(session):
+    homedir = os.getenv("HOME")
+    session.env["SUNDIALS_INST"] = session.env.get("SUNDIALS_INST", f"{homedir}/.local")
+    session.env[
+        "LD_LIBRARY_PATH"
+    ] = f"{homedir}/.local/lib:{session.env.get('LD_LIBRARY_PATH')}"
+    session.install("coverage")
+    session.run("pip", "install", "-e", ".")
+    if sys.platform != "win32" or sys.platform != "darwin":
+        session.install("scikits.odes")
+        session.run("pybamm_install_jax")
+    session.run("coverage", "run", "--rcfile=.coveragerc", "run-tests.py", "--nosub")
+    session.run("coverage", "combine")
+    session.run("coverage", "xml")
+
+
+@nox.session(name="integration", reuse_venv=True)
+def run_integration(session):
+    session.run("pip", "install", "-e", ".")
+    session.run("python", "run-tests.py", "--integration")
+
+
+@nox.session(name="doctests", reuse_venv=True)
+def run_doctests(session):
+    session.run("pip", "install", "-e", ".[docs]")
+    session.run("python", "run-tests.py", "--doctest")
+
+
+@nox.session(name="unit", reuse_venv=True)
+def run_unit(session):
+    session.run("pip", "install", "-e", ".")
+    if sys.platform == "linux":
+        session.run("pybamm_install_jax")
+    session.run("python", "run-tests.py", "--unit")
+
+
+@nox.session(name="examples", reuse_venv=True)
+def run_examples(session):
+    session.run("pip", "install", "-e", ".[dev]")
+    session.run("python", "run-tests.py", "--examples")
+
+
+@nox.session(name="dev", reuse_venv=True)
+def set_dev(session):
+    homedir = os.getenv("HOME")
+    LD_LIBRARY_PATH = f"{homedir}/.local/lib:{session.env.get('LD_LIBRARY_PATH')}"
+    envbindir = session.bin
+    session.run("pip", "install", "-e", ".[dev]")
+    session.install("cmake")
+    session.run(
+        "echo",
+        "export",
+        f"LD_LIBRARY_PATH={LD_LIBRARY_PATH}",
+        ">>",
+        f"{envbindir}/activate",
+    )
```

Comment:
Do we require setting these variables again? What I meant with dev -

We're installing the Pythonic dependencies in every session, which will slow down the CI. We should refactor that installation process in a single function and call that function in the CI before any tests (`nox -s dev`).

Reply:
@Saransh-cpp I had tried doing that more than a couple of times before & just now too but since `isolated_build` is turned on every session is creating seperate `.nox/env/lib/python/site-packages` which means even if I ran a session to install all the deps say `dev` I'd still have to install the deps as the session wouldn't look for libraries in `.nox/dev/lib/python/site-packages` but in `.nox/session_name/lib/python/site-packages`.

Reply:
About setting those variables again, I can globally declare the variables above but global declaration is not recommended.

Reply:
Something left to be addressed in this PR is to update the CHANGELOG, and the documentation: to account for the change to `nox` wherever required and breaking changes in the installation and developer workflow processes (if any)

Reply:
> Something left to be addressed in this PR is to update the CHANGELOG, and the documentation: to account for the change to `nox` wherever required and breaking changes in the installation and developer workflow processes (if any)

On it.

Reply:
Almost forgot the docs! Thanks for the reminder!